middle_end/flambda2/to_cmm/to_cmm_primitive.ml

```diff
@@ -592,12 +592,26 @@ let unary_primitive env res dbg f arg =
       let message = dead_slots_msg dbg [c1; c2] [] in
       let expr, res = C.invalid res ~message in
       None, res, expr)
-  | Project_value_slot { project_from; value_slot; kind = _ } -> (
+  | Project_value_slot { project_from; value_slot; kind } -> (
     match
       value_slot_offset env value_slot, function_slot_offset env project_from
     with
     | Live_value_slot { offset; _ }, Live_function_slot { offset = base; _ } ->
-      None, res, C.get_field_gen Asttypes.Immutable arg (offset - base) dbg
+      let memory_chunk : Cmm.memory_chunk =
+        match Flambda_kind.With_subkind.kind kind with
+        | Value -> Word_val
+        | Naked_number
+            (Naked_int32 | Naked_int64 | Naked_nativeint | Naked_immediate) ->
+          Word_int
+        | Naked_number Naked_float -> Double
+        | Region | Rec_info ->
+          Misc.fatal_errorf "Bad kind %a for Project_value_slot"
+            Flambda_kind.With_subkind.print kind
```

Comment:
This code should probably be in a separate function in `to_cmm_shared` where other similar functions already exist (e.g. functions from flambda kinds to cmm machtypes).

Reply:
Fixed, and now also generates `Word_int` for tagged immediates.